spikeextractors/extractors/cedextractors/cedrecordingextractor.py

```diff
@@ -0,0 +1,143 @@
+from spikeextractors import RecordingExtractor
+from .utils import get_channel_info, get_channel_data
+from spikeextractors.extraction_tools import check_get_traces_args
+
+import numpy as np
+from pathlib import Path
+from typing import Union
+
+try:
+    from sonpy import lib as sp
+
+    HAVE_SONPY = True
+except ImportError:
+    HAVE_SONPY = False
+
+PathType = Union[str, Path, None]
+DtypeType = Union[str, np.dtype, None]
+
+
+class CEDRecordingExtractor(RecordingExtractor):
+    """
+    Extracts electrophysiology recordings from .smrx files.
+    The recording extractor always returns channel IDs starting from 0.
+    The recording data will always be returned in the shape of (num_channels,num_frames).
+
+    Parameters
+    ----------
+    file_path: str
+        Path to the .smrx file to be extracted
+    smrx_ch_inds: list of int
+        List with indexes of valid smrx channels. Does not match necessarily
+        with extractor id.
+    """
+
+    extractor_name = 'CEDRecordingExtractor'
+    installed = HAVE_SONPY  # check at class level if installed or not
+    is_writable = False
+    mode = 'file'
+    installation_mesg = "To use the CED extractor, install sonpy: \n\n pip install sonpy\n\n"  # error message when not installed
+
+    def __init__(self, file_path: PathType, smrx_ch_inds: list):
```

Comment:
@luiztauffer how do you get the `smrx_ch_inds`?

It could be nice to have a static function that returns the channel idxs and names given the path to an smrx file.

Reply:
The smrx file can possibly contain much more than just the raw voltage traces from electrodes.
It doesn’t seem like there is way of automatic inferring which channels are the raw voltage traces and which channels are not. I think we’ll depend on the user input for that

Reply:
But we could have a static function that returns the available channels for an smrx file

Reply:
you mean, something that loops through channels and calls utils.get_channel_info() ?

Reply:
For example, this is what we get when extracting some info for each of the existing channels in the smrx file we're working with:
```
(0, 'RhdD-0', 'mV', <DataType.Adc: 1>)
(1, 'RhdD-1', 'mV', <DataType.Adc: 1>)
(2, 'RhdD-2', 'mV', <DataType.Adc: 1>)
(3, 'RhdD-3', 'mV', <DataType.Adc: 1>)
(4, 'RhdD-4', 'mV', <DataType.Adc: 1>)
(5, 'RhdD-5', 'mV', <DataType.Adc: 1>)
(6, 'RhdD-6', 'mV', <DataType.Adc: 1>)
(7, 'RhdD-7', 'mV', <DataType.Adc: 1>)
(8, 'RhdD-8', 'mV', <DataType.Adc: 1>)
(9, 'RhdD-9', 'mV', <DataType.Adc: 1>)
(10, 'RhdD-10', 'mV', <DataType.Adc: 1>)
(11, 'RhdD-11', 'mV', <DataType.Adc: 1>)
(12, 'RhdD-12', 'mV', <DataType.Adc: 1>)
(13, 'RhdD-13', 'mV', <DataType.Adc: 1>)
(14, 'RhdD-14', 'mV', <DataType.Adc: 1>)
(15, 'RhdD-15', 'mV', <DataType.Adc: 1>)
(16, 'RhdD-16', 'mV', <DataType.Adc: 1>)
(17, 'RhdD-17', 'mV', <DataType.Adc: 1>)
(18, 'RhdD-18', 'mV', <DataType.Adc: 1>)
(19, 'RhdD-19', 'mV', <DataType.Adc: 1>)
(20, 'RhdD-20', 'mV', <DataType.Adc: 1>)
(21, 'RhdD-21', 'mV', <DataType.Adc: 1>)
(22, 'RhdD-22', 'mV', <DataType.Adc: 1>)
(23, 'RhdD-23', 'mV', <DataType.Adc: 1>)
(24, 'RhdD-24', 'mV', <DataType.Adc: 1>)
(25, 'RhdD-25', 'mV', <DataType.Adc: 1>)
(26, 'RhdD-26', 'mV', <DataType.Adc: 1>)
(27, 'RhdD-27', 'mV', <DataType.Adc: 1>)
(28, 'RhdD-28', 'mV', <DataType.Adc: 1>)
Channel 29 is type Off, skipping it...
(30, 'Keyboard', '', <DataType.Marker: 5>)
Channel 31 is type Off, skipping it...
(32, 'RhdD-29', 'mV', <DataType.Adc: 1>)
(33, 'RhdD-30', 'mV', <DataType.Adc: 1>)
(34, 'RhdD-31', 'mV', <DataType.Adc: 1>)
(35, 'RhdD-32', 'mV', <DataType.Adc: 1>)
(36, 'RhdD-33', 'mV', <DataType.Adc: 1>)
(37, 'RhdD-34', 'mV', <DataType.Adc: 1>)
(38, 'RhdD-35', 'mV', <DataType.Adc: 1>)
(39, 'RhdD-36', 'mV', <DataType.Adc: 1>)
(40, 'RhdD-37', 'mV', <DataType.Adc: 1>)
(41, 'RhdD-38', 'mV', <DataType.Adc: 1>)
(42, 'RhdD-39', 'mV', <DataType.Adc: 1>)
(43, 'RhdD-40', 'mV', <DataType.Adc: 1>)
(44, 'RhdD-41', 'mV', <DataType.Adc: 1>)
(45, 'RhdD-42', 'mV', <DataType.Adc: 1>)
(46, 'RhdD-43', 'mV', <DataType.Adc: 1>)
(47, 'RhdD-44', 'mV', <DataType.Adc: 1>)
(48, 'RhdD-45', 'mV', <DataType.Adc: 1>)
(49, 'RhdD-46', 'mV', <DataType.Adc: 1>)
(50, 'RhdD-47', 'mV', <DataType.Adc: 1>)
(51, 'RhdD-48', 'mV', <DataType.Adc: 1>)
(52, 'RhdD-49', 'mV', <DataType.Adc: 1>)
(53, 'RhdD-50', 'mV', <DataType.Adc: 1>)
(54, 'RhdD-51', 'mV', <DataType.Adc: 1>)
(55, 'RhdD-52', 'mV', <DataType.Adc: 1>)
(56, 'RhdD-53', 'mV', <DataType.Adc: 1>)
(57, 'RhdD-54', 'mV', <DataType.Adc: 1>)
(58, 'RhdD-55', 'mV', <DataType.Adc: 1>)
(59, 'RhdD-56', 'mV', <DataType.Adc: 1>)
(60, 'RhdD-57', 'mV', <DataType.Adc: 1>)
(61, 'RhdD-58', 'mV', <DataType.Adc: 1>)
(62, 'RhdD-59', 'mV', <DataType.Adc: 1>)
(63, 'RhdD-60', 'mV', <DataType.Adc: 1>)
(64, 'RhdD-61', 'mV', <DataType.Adc: 1>)
(65, 'RhdD-62', 'mV', <DataType.Adc: 1>)
(66, 'RhdD-63', 'mV', <DataType.Adc: 1>)
(67, 'CED_Mech', 'g', <DataType.Adc: 1>)
(68, 'LFP', 'V', <DataType.Adc: 1>)
Channel 69 is type Off, skipping it...
(70, 'MechTTL', 'V', <DataType.Adc: 1>)
(71, 'MechStim', 'V', <DataType.Adc: 1>)
(72, 'Laser', 'V', <DataType.Adc: 1>)
Channel 73 is type Off, skipping it...
Channel 74 is type Off, skipping it...
Channel 75 is type Off, skipping it...
Channel 76 is type Off, skipping it...
Channel 77 is type Off, skipping it...
Channel 78 is type Off, skipping it...
Channel 79 is type Off, skipping it...
Channel 80 is type Off, skipping it...
Channel 81 is type Off, skipping it...
Channel 82 is type Off, skipping it...
Channel 83 is type Off, skipping it...
Channel 84 is type Off, skipping it...
Channel 85 is type Off, skipping it...
Channel 86 is type Off, skipping it...
Channel 87 is type Off, skipping it...
Channel 88 is type Off, skipping it...
Channel 89 is type Off, skipping it...
Channel 90 is type Off, skipping it...
Channel 91 is type Off, skipping it...
Channel 92 is type Off, skipping it...
Channel 93 is type Off, skipping it...
Channel 94 is type Off, skipping it...
Channel 95 is type Off, skipping it...
Channel 96 is type Off, skipping it...
Channel 97 is type Off, skipping it...
Channel 98 is type Off, skipping it...
Channel 99 is type Off, skipping it...
Channel 100 is type Off, skipping it...
```

Reply:
Yes, so that the user can make the channel list based on it

Reply:
how about d78da82?